README.md
# Aurelia 2 Scaffolding skeleton

[![Build Status](https://travis-ci.com/aurelia/new.svg?branch=master)](https://travis-ci.com/aurelia/new)

_Work In Progress_

The scaffolding repo for Aurelia 2 used by the [makes](https://makes.js.org) tool to create new Aurelia 2 projects.

## Create an Aurelia 2 project

First, ensure that you have Node.js v8.9.0 or above installed on your system. Next, using [npx](https://medium.com/@maybekatz/introducing-npx-an-npm-package-runner-55f7d4bd282b), a tool distributed as part of Node.js, we'll create a new Aurelia 2 app. At a command prompt, run the following command:

```bash
npx makes aurelia
```

This will cause `npx` to download the `makes` tool, along with the `aurelia` scaffold from this repo, which it will use to guide you through creating your project.

## Status

- [x] Aurelia convention support (boilerplate free like Aurelia v1) at bundler level
- [x] Basic app with webpack
- [x] Basic app with dumber (successor of CLI built-in bundler. Doc WIP)
- [ ] Basic app with parcel (On hold)
  * Parcel can be done. But requires a dedicated repo (out of our mono repo) in order to write in commonjs format that Parcel wants.
  * https://github.com/parcel-bundler/parcel/issues/3256
  * Parcel v2 is also very near. Might wait for the new version.
- [ ] Basic app with browserify (On hold)
  * Browserify has technical obstacle on TypeScript (tsify) https://github.com/TypeStrong/tsify/issues/34#issuecomment-514425682
  * There is no problem on babelify. But we will hold browserify for now.
- [ ] Basic app with FuseBox (On hold)
  * FuseBox v4 is coming, and it's a total rewrite. Wait for it before implement.
- [x] Basic less/scss setup.
- [x] Basic unit test setup for jest, jasmine, mocha, tape. (ava is on hold, pending implementing transformer.)
- [x] Basic e2e test setup for cypress
- [ ] Basic e2e test setup for protractor

## Development

This scaffolding skeleton is in very early stage, not quite ready for adding features. We want to keep feature set manageable as Aurelia 2 is constantly evolving.

There are some tests for this skeleton, setup in package.json. (totally not required by makes)

```bash
npm install
# test some top level makes hooks
npm test
# test all skeletons, take some time, not turned on in .travis.yml
npm run test:e2e
```

To target subset of skeletons, use env variable `TARGET_FEATURES`.

```bash
# only test skeletons using webpack and typescript features.
npx cross-env TARGET_FEATURES=webpack,typescript npm run test:e2e
```

If you forked this repo, you can run your skeleton with:

```bash
# test your master branch if your forked name is "new"
npx makes your_GitHub_name
# test some branch or commit or tag
npx makes your_GitHub_name/forked_repo_name#some-branch
```

## License

MIT.


CHANGELOG.md
# [0.2.0](https://github.com/aurelia/new/compare/v0.1.0...v0.2.0) (2020-04-16)


### Bug Fixes

* **app-min:** skip test on no-unit-tests ([85bc12a](https://github.com/aurelia/new/commit/85bc12adb00305a9fe1d56d42ac62a3af76e020d))
* **app-with-router:** set .goto-active as global css class when using css-module ([cd0375f](https://github.com/aurelia/new/commit/cd0375fa94a211b5661f06b943787cadb94eb533)), closes [aurelia/aurelia#828](https://github.com/aurelia/aurelia/issues/828)
* **dumber:** cleanup tests, fix css-module tests ([986cc26](https://github.com/aurelia/new/commit/986cc26251144207ac4fc8f9293f436745713b89))
* **tape:** fix tape version ([5eb122a](https://github.com/aurelia/new/commit/5eb122a4a46fabc2b183027202dcf64f30fea9f9))
* **tape:** fix typescript import ([a0d0fef](https://github.com/aurelia/new/commit/a0d0fef167ce0ce55949802c6c749829dacd8739))
* **test:** fix outdated test code for shadow-dom and css-modules ([f111b54](https://github.com/aurelia/new/commit/f111b54a0a20485c465152fa34a10384f92a5d1b))
* **test:** fix setup js/ts code ([5273d3f](https://github.com/aurelia/new/commit/5273d3f704042f84ac59ea4a31e9825801be4558))
* **test:** scheduler has been moved ([78770e2](https://github.com/aurelia/new/commit/78770e2b09790842130872808c15d36e7ad09098)), closes [aurelia/aurelia#831](https://github.com/aurelia/aurelia/issues/831)
* **typescript:** add missing module def for less and scss files ([9947f33](https://github.com/aurelia/new/commit/9947f339305302754eb4604dd007946e14613b75))
* **webpack:** allow shared style import in main file for setup with shadowDom ([fcc8a5c](https://github.com/aurelia/new/commit/fcc8a5cf0ccce1b5637990cf1106703766a569dc))
* **webpack:** bypass tsc for test setup ([b24d71f](https://github.com/aurelia/new/commit/b24d71f7b0d7f11824482b5e961657c65cbae046))
* **webpack:** cleanup duplicated loader ([5b2f88a](https://github.com/aurelia/new/commit/5b2f88aa2ee205a1ea94e7d4fe1c16d5e3f2bb89))
* **webpack:** fix css setup on resources ([4463813](https://github.com/aurelia/new/commit/446381399a708458ad726068e5f6c88955c91a10)), closes [#18](https://github.com/aurelia/new/issues/18)
* **webpack:** fix missing analyze ([b22389c](https://github.com/aurelia/new/commit/b22389c9abe07ca04079aa2c9ae4b3f66d794daa))
* **webpack:** fix shared shadow-dom styles ([cdd50fd](https://github.com/aurelia/new/commit/cdd50fd63415dcaa682eb0c7604f57eb341a3f1f))
* bypass style-loader when shadow-dom is in use ([160eb61](https://github.com/aurelia/new/commit/160eb61c149769db2fc238efea5dac194736fa08))
* CustomElement.behaviorFor has been changed to CustomElement.for ([bd2b16a](https://github.com/aurelia/new/commit/bd2b16a061e625df9bc8f1cd7aff658214a3dcc4))
* fix dumber gulp file ([071e5e3](https://github.com/aurelia/new/commit/071e5e3866d883c465b7109add2dddb63adfde9e))
* fix dumber gulpfile on test build ([f30163b](https://github.com/aurelia/new/commit/f30163b0038e9a9e90b819464ae8766fae932335))
* fix sass-lint version ([5466b50](https://github.com/aurelia/new/commit/5466b507393d5ce99d0c5293d8f0dd7e0d208fd0))
* fix sass-loader options ([1f030d1](https://github.com/aurelia/new/commit/1f030d12d529fd9802c5bd9dbb58725bc59b51e9))
* fix shadowRoot typing in TS ([b55a327](https://github.com/aurelia/new/commit/b55a3272003c24d1cecd2c5429f80455a5e4dbdb))
* fix sourcemap of dumber browser test ([61b7393](https://github.com/aurelia/new/commit/61b73936cca07fc5886f0f9e235ba394fc6b5794))
* fix tape usage on fs ([10c0dbd](https://github.com/aurelia/new/commit/10c0dbd3f0e3a214e9d4df5221430cf3d4f2f906))
* fix ts typing in unit tests ([1051588](https://github.com/aurelia/new/commit/10515889eba3dfbed429a589d9de7b1da0e181dc))
* fix tslint error ([07f52d2](https://github.com/aurelia/new/commit/07f52d2ebec6439e81e0466db3cd5bedee461409))
* fix webpack test setup ([9c63cbd](https://github.com/aurelia/new/commit/9c63cbd37fb307561dace21eb940deb33ad881c5))
* fix webpack watch mode unit test ([3bc28aa](https://github.com/aurelia/new/commit/3bc28aa23c49e62cbb32e0394133cd86be07b9e7))
* should reject error in test mode ([022b6bb](https://github.com/aurelia/new/commit/022b6bb460687ac2d6a296cc082fa314ef616247))
* should use absolute src path in webpack resolve.modules ([1e1f80a](https://github.com/aurelia/new/commit/1e1f80a26a2ea67ea7b2742cf602225a9716883e))
* use to-string-loader to support shadow-dom css ([28a8136](https://github.com/aurelia/new/commit/28a8136044c87088d3be2e7f6291a60f1676d127))
* **webpack:** missing path import in webpack config ([e23ebdc](https://github.com/aurelia/new/commit/e23ebdce916f0a4758008273be561191a9db3f46))


### Features

* use sass rather than node-sass ([#16](https://github.com/aurelia/new/issues/16)) ([51ffd74](https://github.com/aurelia/new/commit/51ffd747c80470ba12d0fcb8efd9493bea503c52))
* **all:** add option for app-with-router ([327f67d](https://github.com/aurelia/new/commit/327f67d536263306f2438fa7d5aea0e90a13d9c1))
* **test:** proper test setup ([a5896af](https://github.com/aurelia/new/commit/a5896afdfa2b63fe2ca7c126c3a4c25ae7fe9fbd)), closes [#11](https://github.com/aurelia/new/issues/11)
* **webpack:** support --analyze ([2b3a9a1](https://github.com/aurelia/new/commit/2b3a9a1331c7138a4bd73e8c7661c919f618fc3f))
* add options of less/sass ([8277f4e](https://github.com/aurelia/new/commit/8277f4e1b2b6d69398f558137adc02f6369a6daa))
* add sass/less linter ([1df677a](https://github.com/aurelia/new/commit/1df677a446eec5301f7cc6fb2828518349f72535))
* support dynamic import in babel ([97ada26](https://github.com/aurelia/new/commit/97ada261f44024044a705503c9e7e5fc374d0666))
* **webpack:** switch webpack config to use env string ([82bd63f](https://github.com/aurelia/new/commit/82bd63f93b024c1d774c930f6fe71fa0fd64490d))


### BREAKING CHANGES

* **webpack:** webpack env is now controlled by --env production or NODE_ENV.



# 0.1.0 (2019-10-02)


### Bug Fixes


questions.js
module.exports = [
  {
    message: 'Which bundler would you like to use?',
    choices: [
      {value: 'webpack', title: 'Webpack', hint: 'A powerful and popular bundler for modern JavaScript apps.'},
      {value: 'dumber', title: 'Dumber', hint: 'A dumb JavasScript bundler, dumber than you and me. The successor of Aurelia CLI built-in bundler.'}

      // Pending implementation of conventions plugins for these bundlers.

      // Parcel can be done. But requires a dedicated repo (out of our mono repo) in order to write in commonjs format that Parcel wants.
      // https://github.com/parcel-bundler/parcel/issues/3256
      // Parcel v2 is also very near. Might wait for the new version.
      // {value: 'parcel', title: 'Parcel', hint: 'Blazing fast, zero configuration web application bundler.'},

      // Browserify has technical obstacle on TypeScript (tsify) https://github.com/TypeStrong/tsify/issues/34#issuecomment-514425682
      // There is no problem on babelify. But we will hold browserify for now.
      // {value: 'browserify', title: 'Browserify', hint: 'The bundler who started the business, a tool for compiling node-flavored commonjs modules for the browser.'},

      // FuseBox v4 is coming, and it's a total rewrite. Wait for it before implement.
      // {value: 'fuse-box', title: 'FuseBox', hint: 'A bundler that does it all, and does it right.'},
    ]
  },
  {
    message: 'What transpiler would you like to use?',
    choices: [
      {value: 'babel', title: 'Babel', hint: 'An open source, standards-compliant ES2015 and ESNext transpiler.'},
      {value: 'typescript', title: 'TypeScript', hint: 'An open source, ESNext superset that adds optional strong typing.'}
    ]
  },
  {
    message: 'Do you want to use Shadow DOM?',
    choices: [
      {title: 'No'},
      {value: 'shadow-dom', title: 'Yes', hint: 'Shadow DOM in open mode, shadowRoot is accessible through DOM API.'},
    ]
  },
  {
    // Don't allow user to turn on both ShadowDOM and CSS Module.
    if: '!shadow-dom',
    message: 'Do you want to use CSS Module?',
    choices: [
      {title: 'No'},
      {value: 'css-module', title: 'Yes', hint: 'CSS Module is an alternative way to locally scope CSS class names. Do not use it together with Shadow DOM.'},
    ]
  },
  {
    message: 'What CSS preprocessor to use?',
    choices: [
      {value: 'css', title: 'Plain CSS'},
      {value: 'less', title: 'Less (.less)'},
      {value: 'sass', title: 'Sass (.scss)'}
    ]
  },
  {
    message: 'What unit testing framework to use?',
    choices: [
      {value: 'jest', title: 'Jest', hint: 'Runs in Node.js, simulates browser by default, with a focus on simplicity.'},
      // Disable ava for now.
      // TODO: build some custom transformer to apply our conventions when running in nodejs env.
      // {value: 'ava', title: 'Ava + jsdom-global', hint: 'Runs in Node.js, simulates browser with jsdom-global. A test runner for Node.js with a concise API, detailed error output, embrace of new language features and process isolation that let you write tests more effectively.'},
      {value: 'jasmine', title: 'Jasmine', hint: 'Runs in browser, a behavior-driven testing framework.'},
      {value: 'mocha', title: 'Mocha + Chai', hint: 'Runs in browser, a feature-rich JavaScript test framework for node and browsers.'},
      {value: 'tape', title: 'Tape', hint: 'Runs in browser, tap-producing test harness for node and browsers.'},
      {value: 'no-unit-tests', title: 'None', hint: 'Skip testing. My code is always perfect :-)'}
    ]
  },
  {
    message: 'Do you want to setup e2e test?',
    choices: [
      {title: 'No'},
      {value: 'cypress', title: 'Yes (Cypress)', hint: 'Cypress offers fast, easy and reliable testing for anything that runs in a browser.'}
    ]
  },
  {
    message: 'What kind of sample code do you want in this project?',
    choices: [
      {value: 'app-min', title: 'Bare minimum'},
      {value: 'app-with-router', title: 'With direct routing'}
    ]
  },
  // Currently this question is not visible to end user because there is only one option.
  // Will expand the list in future.
  {
    message: 'What are your targeted browsers?',
    choices: [
      {value: 'browser-evergreen', title: 'All Evergreen browsers', hint: 'Including Chrome, Edge, Firefox, Safari'}
    ]
  }
];
